eth/tracers/native/contract.go

```
@@ -0,0 +1,134 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package native
+
+import (
+	"encoding/json"
+	"math/big"
+	"sync/atomic"
+	"time"
+
+	"github.com/XinFinOrg/XDPoSChain/common"
+	"github.com/XinFinOrg/XDPoSChain/core/vm"
+	"github.com/XinFinOrg/XDPoSChain/eth/tracers"
+)
+
+func init() {
+	tracers.RegisterNativeTracer("contractTracer", NewContractTracer)
+}
+
+type contractTracer struct {
+	Addrs     map[string]string
+	config    contractTracerConfig
+	interrupt uint32 // Atomic flag to signal execution interruption
+	reason    error  // Textual reason for the interruption
+}
+
+type contractTracerConfig struct {
+	OpCode       string `json:"opCode"`       // Target opcode to trace
+	WithByteCode bool   `json:"withByteCode"` // If true, bytecode will be collected
+}
+
+// NewContractTracer returns a native go tracer which tracks the contracr was created
+func NewContractTracer(cfg json.RawMessage) (tracers.Tracer, error) {
+	var config contractTracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, err
+		}
+	}
+	t := &contractTracer{
+		Addrs:  make(map[string]string, 1),
+		config: config,
+	}
+	// handle invalid opcode case
+	op := vm.StringToOp(t.config.OpCode)
+	if op == 0 && t.config.OpCode != "STOP" && t.config.OpCode != "" {
+		t.config.OpCode = "inv"
```

I would return an error here instead. No reason to do tracing in this case.